kornia/filters/guided.py

```diff
@@ -0,0 +1,217 @@
+        cov_Ip = corr_Ip.reshape(B, H, W, C, -1) - mean_p.unsqueeze(-2) * mean_I.unsqueeze(-1)
+
+    if isinstance(eps, Tensor):
+        _eps = torch.eye(C, device=guidance.device, dtype=guidance.dtype).view(1, 1, 1, C, C) * eps.view(-1, 1, 1, 1, 1)
+    else:
+        _eps = guidance.new_full((C,), eps).diag().view(1, 1, 1, C, C)
+    a = torch.linalg.solve(var_I + _eps, cov_Ip)  # B, H, W, C_guidance, C_input
+    b = mean_p - (mean_I.unsqueeze(-2) @ a).squeeze(-2)  # B, H, W, C_input
+
+    mean_a = box_blur(a.flatten(-2).permute(0, 3, 1, 2), kernel_size, border_type)
+    mean_b = box_blur(b.permute(0, 3, 1, 2), kernel_size, border_type)
+
+    if subsample > 1:
+        mean_a = interpolate(mean_a, scale_factor=subsample, mode="bilinear")
+        mean_b = interpolate(mean_b, scale_factor=subsample, mode="bilinear")
+    mean_a = mean_a.view(B, C, -1, H * subsample, W * subsample)
+
+    # einsum might not be contiguous, thus mean_b is the first argument
+    return mean_b + torch.einsum("BCHW,BCcHW->BcHW", guidance, mean_a)
+
+
+def guided_blur(
+    guidance: Tensor,
+    input: Tensor,
+    kernel_size: tuple[int, int] | int,
+    eps: float | Tensor,
+    border_type: str = 'reflect',
+    subsample: int = 1,
+) -> Tensor:
+    r"""Blur a tensor using a Guided filter.
+
+    .. image:: _static/img/guided_blur.png
+
+    The operator is an edge-preserving image smoothing filter. See :cite:`he2010guided`
+    and :cite:`he2015fast` for details. Guidance and input can have different number of channels.
+
+    Arguments:
+        guidance: the guidance tensor with shape :math:`(B,C,H,W)`.
+        input: the input tensor with shape :math:`(B,C,H,W)`.
+        kernel_size: the size of the kernel.
+        eps: regularization parameter. Smaller values preserve more edges.
+        border_type: the padding mode to be applied before convolving.
+          The expected modes are: ``'constant'``, ``'reflect'``,
+          ``'replicate'`` or ``'circular'``. Default: ``'reflect'``.
+        subsample: subsampling factor for Fast Guided filtering. Default: 1 (no subsampling)
+
+    Returns:
+        the blurred tensor with same shape as `input` :math:`(B, C, H, W)`.
+
+    Examples:
+        >>> guidance = torch.rand(2, 3, 5, 5)
+        >>> input = torch.rand(2, 4, 5, 5)
+        >>> output = guided_blur(guidance, input, 3, 0.1)
+        >>> output.shape
+        torch.Size([2, 4, 5, 5])
+    """
+
+    KORNIA_CHECK_IS_TENSOR(guidance)
+    KORNIA_CHECK_SHAPE(guidance, ['B', 'C', 'H', 'W'])
+    if input is not guidance:
+        KORNIA_CHECK_IS_TENSOR(input)
+        KORNIA_CHECK_SHAPE(input, ['B', 'C', 'H', 'W'])
+        KORNIA_CHECK(
+            (guidance.shape[0] == input.shape[0]) and (guidance.shape[-2:] == input.shape[-2:]),
+            "guidance and input should have the same batch size and spatial dimensions",
+        )
+
+    kernel_size = _unpack_2d_ks(kernel_size)
+    if guidance.shape[1] == 1:
```

Comment:
we should add raise as optional in KORNIA_CHECK so we can use it like:

```python
if KORNIA_CHECK_IS_GRAY(guidance, raise=False):
    ...
```
cc\ @edgarriba

Reply:
i like it /cc @ducha-aiki